### ISSUE TITLE
Preventing unmanaged Cilium endpoints on newly-created nodes

### ISSUE THREAD

**Issue by ti-mo** [cilium/cilium#16602]
## Proposal / RFE

See https://github.com/cilium/cilium/issues/15366 for one recent report of this.

**Problem**: when a nodepool scale-out event happens (often when there's a certain amount of unscheduled Pods waiting for resources), workloads are often scheduled on new nodes before the Cilium agent has a chance to pull and run. The result is Cilium-unmanaged endpoints. The agent image is several hundreds of MB large, so this occurs frequently enough to be noticeable, especially with some of our users that run at scale.

We've been discussing solutions to this internally for a while, but one of the working proposals is as follows:

Kubelet has the `--register-with-taints` parameter to taint its own node resource atomically when it first registers itself to the Kubernetes API. There is no window of time where the node can exist without these taints, so it's guaranteed to block scheduling (unless a Pod tolerates all taints, which we can't/won't account for). Notably, this also blocks `DaemonSets`, since those tolerate [only a specific list of taints](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#taints-and-tolerations) Once the Cilium agent has started up successfully, it removes the specified taint, at which point workload scheduling can resume.

Practically, this means:

- During cluster or nodepool creation, the user specifies a Cilium-specific on-register taint, let's say `node.cilium.io/agent-not-ready`. This will ensure the taint is automatically applied to all newly-created nodes.
  - Azure AKS: `az aks nodepool add .. --node-taints node.cilium.io/agent-not-ready=true:NoSchedule` ([docs](https://docs.microsoft.com/en-us/cli/azure/aks/nodepool?view=azure-cli-latest#az_aks_nodepool_add))
    **Note**: AKS only allows specifying taints on specific node pools, not on the cluster as a whole. This needs to be taken into account during cluster planning. Since autoscaling nodepools are commonly run separately, this would not be a blocker, and existing clusters can be provisioned with new nodepools to take advantage of this feature when released.
  - Google GKE: `gcloud container clusters create cluster-name --node-taints node.cilium.io/agent-not-ready=true:NoSchedule` ([docs](https://cloud.google.com/kubernetes-engine/docs/how-to/node-taints#creating_a_cluster_with_node_taints)) - can be set on either a cluster or nodepool level!
  - Amazon EKS: depends on the provisioning method, but bootstrap.sh has support for arbitrary kubelet arguments through `--kubelet-extra-args` ([docs](https://aws.amazon.com/blogs/opensource/improvements-eks-worker-node-provisioning)) and `eksctl` supports setting taints on `nodeGroups` ([docs](https://eksctl.io/usage/autoscaling/#scaling-up-from-0)).
- The Cilium agent DaemonSet would tolerate this taint, so it would be allowed to schedule first. Once the agent has started successfully, it removes the taint from its own node resource.
- All other Pods will now get the opportunity to schedule.
- (see below) On shutdown, reapply `node.cilium.io/agent-not-ready=true:NoSchedule` to the local node.
  @aanm This would be another angle to the problem `*.cilium_bak` tries to solve.

**Caveat**: Taints are not reapplied when Kubelet restarts, when the node reboots, or when the Cilium agent is restarted. In case of a full node reboot, or a Cilium agent restart, this still leaves a window for Pods to be scheduled without receiving Cilium identities. To combat this, the Cilium agent could taint its own node during shutdown, which would prevent workloads from being scheduled while the Cilium agent is down (including during early kubelet startup).

cc @bmcustodio @joestringer

**Comment by aanm** [comment]
This sounds like a great proposal. The only issue I see with it is that hostNetwork pods would not be able to be scheduled on any node but in this case it would just be a matter of adding the toleration for `node.cilium.io/agent-not-ready` in such pods.

**Comment by joestringer** [comment]
Q: How does `cilium-operator` intersect with this proposal? Is it a component we should mark as tolerating this taint?

> Caveat: Taints are not reapplied when Kubelet restarts, when the node reboots, or when the Cilium agent is restarted. In case of a full node reboot, or a Cilium agent restart, this still leaves a window for Pods to be scheduled without receiving Cilium identities. To combat this, the Cilium agent could taint its own node during shutdown, which would prevent workloads from being scheduled while the Cilium agent is down (including during early kubelet startup).

Tainting nodes during Cilium shutdown is an interesting idea, seems like it has the possibility to help solve some of this, not 100% of cases but maybe good enough. This particular caveat seems like something that requires upstream k8s discussion to sort the model out properly and clarify what CNIs _should_ be doing to address these issues.

**Comment by sharkymcdongles** [comment]
I think we need this on GKE. It seems if the CNI doesn't boot fast enough on the new Dataplane V2 clusters then the folder fills up with IPs that are unmanaged and not actually assigned:

`/var/lib/cni/networks/gke-pod-network`

leading to this error despite having 2 full /18 ranges(1 for services 1 for pods) to assign IPs from (over 32k):
```
Events:
  Type     Reason                  Age                   From               Message
  ----     ------                  ----                  ----               -------
  Warning  FailedScheduling        42m (x6 over 43m)     default-scheduler  0/26 nodes are available: 23 node(s) didn't match Pod's node affinity, 3 node(s) were unschedulable.
  Warning  FailedScheduling        42m                   default-scheduler  0/29 nodes are available: 23 node(s) didn't match Pod's node affinity, 3 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate, 3 node(s) were unschedulable.
  Normal   Scheduled               42m                   default-scheduler  Successfully assigned ingress-nginx/ingress-nginx-controller to gke-mip-ppr-clone-5yd1-preemptible-5f23af66-uqc6
  Warning  FailedCreatePodSandBox  42m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "93258238fc683575d6a1b1d80ec61d220ad0756dc5fbe4917da6023bdc56c939": failed to find plugin "cilium-cni" in path [/home/kubernetes/bin]
  Warning  FailedCreatePodSandBox  42m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "014c4c69028fd235db1c9b2777fadf1ec1d317906ca773eaa73bf503c07c5563": failed to find plugin "cilium-cni" in path [/home/kubernetes/bin]
  Warning  FailedCreatePodSandBox  41m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "16fc5a60b139b0cedef8953369d66590cdda9688fa3a9520db14787e09fb75c5": failed to allocate for range 0: no IP addresses available in range set: 10.71.14.1-10.71.14.254
  Warning  FailedCreatePodSandBox  41m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "05ea5a69d562f6c968ea97c30330e493524921721c6cffea023f2f92d0ba5bf4": failed to allocate for range 0: no IP addresses available in range set: 10.71.14.1-10.71.14.254
  Warning  FailedCreatePodSandBox  41m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "b2441cbd303c8f9306c5ad55051fc492fd5f71974bed04be7ba8be5bfa079115": failed to allocate for range 0: no IP addresses available in range set: 10.71.14.1-10.71.14.254
  Warning  FailedCreatePodSandBox  41m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "74d0daac8c3ee78430dee2a2b73fb5ece37b982e82d25c34b5a72ece2cc8b38e": failed to allocate for range 0: no IP addresses available in range set: 10.71.14.1-10.71.14.254
  Warning  FailedCreatePodSandBox  40m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "61bbd3f814238d89e2728825ad74c7e525c7790645ee97ca8ff350124a69eb1c": failed to allocate for range 0: no IP addresses available in range set: 10.71.14.1-10.71.14.254
  Warning  FailedCreatePodSandBox  40m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "4b840025120a29c1293a42d27857c6e5465938b2b34dc37cc697443b5cffe42e": failed to allocate for range 0: no IP addresses available in range set: 10.71.14.1-10.71.14.254
  Warning  FailedCreatePodSandBox  40m                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "2f1c4d86157e6ff7c87e74ab564dcfa6fcab8b0f30e3f23433321028122d0468": failed to allocate for range 0: no IP addresses available in range set: 10.71.14.1-10.71.14.254
  Warning  FailedCreatePodSandBox  118s (x167 over 39m)  kubelet            (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "9f2f57046d6f09ea0898b8c684e90f6923671fd294499d7aa2cb8bd0bbba253e": failed to allocate for range 0: no IP addresses available in range set: 10.71.14.1-10.71.14.254
```
